xla/backends/profiler/gpu/cupti_tracer.h

```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+
+  // Requires pm sampler object
+  absl::Status StopSampling();
+
+  // Requires pm sampler object
+  absl::Status DisableSampling();
+
+  // Collect sampling data
+  // Requires pm sampler object, counter data image, fetches data from hw
+  // buffer into counter data image
+  absl::Status FillCounterDataImage(struct PmSamplingDecodeInfo& decodeInfo);
+
+  // Requires counter data image
+  absl::Status GetSampleCounts(struct PmSamplingDecodeInfo& decodeInfo);
+
+  // Requires host object, pm sampler object, counter data image, metric
+  // names, returns sample time, metric values
+  absl::Status GetSample(SamplerRange& sample, size_t index);
+
+  // Requires pm sampler object, counter data image, (re)initializes it
+  absl::Status InitializeCounterDataImage();
+
+  // Restores image from backup (faster than re-initializing)
+  absl::Status RestoreCounterDataImage();
+
+  // Constructor provides all configuration needed to set up sampling on a
+  // single device
+  PmSamplingDevice(int device_id, struct PmSamplingConfig* config);
+};
+
+// Container for PM sampling decode thread
+// Responsible for fetching PM sampling data from device and providing it to
+// handler or other external container
+class PmSamplingDecodeThread {
+  enum thd_state {
+    // Thread is starting, not yet ready to be enabled
+    state_uninitialized_,
+    // Thread is ready for enablement but decoding has not yet been triggered
+    state_initialized_,
+    // Thread is enabled, polling for metrics from all devices
+    state_enabled_,
+    // Thread is disabled, not polling for metrics, but could be re-enabled
+    state_disabled_,
+    // Thread is finishing and guaranteed to return, allowing join
+    state_exiting_
+  };
+
+  // Current state of the thread (only set by worker thread)
+  volatile thd_state current_thd_state_ = state_uninitialized_;
+    
+  // State thread should transition to (only set by main thread)
+  volatile thd_state nextThdState_ = state_initialized_;
+
+  // Thread handle
+  std::thread* thd_;
+
+  // Function run by thd_
+  static void ThdFunc(PmSamplingDecodeThread* control);
+
+  // Isolate the main decode loop
+  static void ThdFuncDecodeUntilDisabled(PmSamplingDecodeThread* control);
+
+  // Devices to decode by this thread
+  std::vector< std::shared_ptr<PmSamplingDevice> > devs_;
+
+  public:
+  // Spin wait sleep period, set to the min of this and all device periods
+  // Space to asynchronously initialize this class and the thread it spawns
+  absl::Duration decode_period_ = absl::Seconds(1);
+  PmSamplingDecodeThread(std::vector< std::shared_ptr<PmSamplingDevice> >
+      devs);
+
+  // Signal thread to exit; join thread
+  ~PmSamplingDecodeThread() {
+    nextThdState_ = state_exiting_;
+    thd_->join();
+  }
+
+  // Signal and test for state transitions
+  bool IsThdInitialized() { return current_thd_state_ == state_initialized_; }
+  void ThdIsInitialized() { current_thd_state_ = state_initialized_; }
```

Comment:
I find this function name and its siblings confusing. ThdIsInitialized sounds like a boolean predicate, not a method that mutates the class state.